mypy/checkexpr.py

```diff
@@ -5082,15 +5087,27 @@ def fast_container_type(
                 # fallback to slow path
                 self.resolved_type[e] = NoneType()
                 return None
-            values.append(self.accept(item))
-        vt = join.join_type_list(values)
-        if not allow_fast_container_literal(vt):
+
+            typ = self.accept(item)
+            if typ not in values:
+                values.append(typ)
+
+        vt = self._first_or_join_fast_item(values)
+        if vt is None:
             self.resolved_type[e] = NoneType()
             return None
         ct = self.chk.named_generic_type(container_fullname, [vt])
         self.resolved_type[e] = ct
         return ct
 
+    def _first_or_join_fast_item(self, items: list[Type]) -> Type | None:
+        if len(items) == 1 and not self.chk.current_node_deferred:
+            return items[0]
+        typ = join.join_type_list(items)
+        if not allow_fast_container_literal(typ):
```

Comment:
This feels a bit restrictive together with the above `if`. For example, if there is more than one callable type (like in the original example in the issue) this will not apply. Should we also proceed with fast path if each expression is a `RefExpr` (so that its type doesn't depend on the context).

Reply:
We could, but our join implementation for callables is horribly broken, so this will produce invalid results: e.g. we can confidently join two callables with named args with different names into one of them.

```python
def left(*, x: str): ...
def right(*, y: str): ...
```

then `join(left, right)` is `def (*, y: str)`.

I'm not sure if there are other similarly terrifying bugs there, so I'd really prefer to not expand this logic now. This is not related to expression source (`[left, right]` contains to `RefExpr`s). I added a relevant todo here.